vault/quotas/quotas_rate_limit.go

```diff
@@ -181,24 +255,59 @@ func (rlq *RateLimitQuota) allow(req *Request) (Response, error) {
 		return resp, fmt.Errorf("missing request client address in quota request")
 	}
 
-	limit, remaining, reset, allow := rlq.store.Take(req.ClientAddress)
-	resp.Allowed = allow
-	resp.Headers = map[string]string{
-		httplimit.HeaderRateLimitLimit:     strconv.FormatUint(limit, 10),
-		httplimit.HeaderRateLimitRemaining: strconv.FormatUint(remaining, 10),
-		httplimit.HeaderRateLimitReset:     time.Unix(0, int64(reset)).UTC().Format(time.RFC1123),
+	var retryAfter string
+
+	defer func() {
+		if !resp.Allowed {
+			resp.Headers[httplimit.HeaderRetryAfter] = retryAfter
+			rlq.metricSink.IncrCounterWithLabels([]string{"quota", "rate_limit", "violation"}, 1, []metrics.Label{{"name", rlq.Name}})
+		}
+	}()
+
+	rlq.lock.Lock()
```

Comment:
I think it's important to make this a read lock instead of a write lock, because quite a lot of different requests could be contending on it.  But then you have to acquire a write-lock to modify rlq.blockedClients.  Making it a sync.Map instead would remove that as a problem.

Reply:
So after making those tweaks, `sync.Map` and the tweaked version perform nearly identical in benchmarks (1000 unique clients for executed for 1m). Being that `sync.Map` requires less LOC, I'll go that route.